🐛 Source Shopify: fixed `store` redirection, add `bulk checkpointing`

bazarnov opened the pull request:
## What
Resolving:
- https://github.com/airbytehq/airbyte/issues/39499
- https://github.com/airbytehq/oncall/issues/5850
- https://github.com/airbytehq/oncall/issues/5866

## How
- added new transient `Exception` --> `BulkJobRedirectToOtherShopError,` which occurs when the store is operated from the manager account and the responses are redirected to the original `shop` instead of the `aliased` one
- removed `session` and `stream_name` input arguments for `job_manager` for BULK streams, in favor of the `http_client` setup
- added checkpointing ability to all `BULK` streams:
   - added new `config` field --> `job_checkpoint_interval` with the default value of `200,000` (200k lines) to be collected, before the single Bulk Job is checkpointed. 
   - once there are more than the `threshold` lines collected for the single BULK Job, it's `Canceled` and the intermediate results are processed, instead of complete cancelation and retry using smaller `job_size`, which is also there, but will be used as a `safety net` for `long-running` jobs. The primary method to overcome `data-density` issues for heavy-data-loaded streams is `checkpointing`.
- added/updated `unit tests` to cover the updates


The checkpointing for the Shopify Bulk Operation looks like this:
- Bulk Job is running
- Bulk Job satisfies the condition of `job_should_checkpoint` (once the server has already collected some results, based on the `objectCount` field, within the `response` for the `RUNNING` job)
- Bulk Job is `CANCELED`
- Intermediate results are processed, due to the `URL` provided for the already collected results, and the fact the Job is canceled
- The intermediate results are processed and the checkpoint is saved for each record observed.
- The next slice will `start` from the `checkpointed` cursor value, the `end` is also adjusted according to the `bulk_window_in_days` step.

## User Impact
No impact is expected, except for the fact the stuck syncs that were not able to make any progress due to `Canceled / Retried` Bulk Jobs should be fetched normally, and the progress should take place now.

## Can this PR be safely reverted and rolled back?
- [X] YES 💚
- [ ] NO ❌

bazarnov commented:
The new `pre-release` tag was published: `2.4.14-dev.f03bf1112`.

This release contains:
- store redirection to the original store
- adds bulk checkpointing at the `200k` records (adjusted in the config)

brianjlai commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:
```diff
@@ -206,23 +220,26 @@ def _log_job_state_with_count(self) -> None:
 
     def _log_state(self, message: Optional[str] = None) -> None:
-        pattern = f"Stream: `{self.stream_name}`, the BULK Job: `{self._job_id}` is {self._job_state}"
+        pattern = f"Stream: `{self.http_client._name}`, the BULK Job: `{self._job_id}` is {self._job_state}"
```
A minor complaint, but I think we should exercise rigor and try to avoid referencing CDK component's private fields. In this case the HttpClient's private '_name' field. We do this in a few places

It's a bit tedious to add back the stream_name as a init param, and this case is somewhat trivial, but I think we should avoid this habit. Exposing the http client's name is something we probably should have done before and we can do now, but if you don't want to block on another CDK fix

bazarnov replied:
It's very handy to have this inside of http_client, since we don't need to provide more `dummy` attrs to the instances of something like `job_manager`, but just pass the http_client with all the information pre-defined. 

So you recommend not to use this awesome feature, but what are the risks here you see?

bazarnov replied:
Updated the cdk to `3.9.3`, which includes the `name` property, and changed the references. Works as expected. Thanks!

brianjlai commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:
```diff
@@ -234,8 +251,16 @@ def _job_get_result(self, response: Optional[requests.Response] = None) -> Optio
                 self._log_job_state_with_count()
+            elif self._job_state in [ShopifyBulkJobStatus.CANCELED.value, ShopifyBulkJobStatus.CANCELING.value]:
+                # do not emit `CANCELED / CANCELING` Bulk Job status, while checkpointing
+                if self._job_should_checkpoint:
```
should this instead be:
```
if not self._job_should_checkpoint:
  self._log_job_state_with_count()
```

bazarnov replied:
Corrected. Thanks.

brianjlai commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:
```diff
@@ -245,19 +270,36 @@ def _on_created_job(self, **kwargs) -> None:
+        self.logger.info(
+            f"Stream: `{self.http_client._name}` the BULK Job: {self._job_id} runs longer than expected ({self._job_max_elapsed_time} sec). Retry with the reduced `Slice Size` after self-cancelation."
+        )
+        self._job_long_running_cancelation = True
```
We set this field to True, but as far as I can tell, we don't actually read or use this value. Do we still need it outside of assertions on tests? And if this is purely to validate tests, let's comment in the above field declaration that it's included for tests

bazarnov replied:
Removed! That was a leftover from couple of previous test attempts. Thanks!

brianjlai commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:
```diff
@@ -339,6 +385,32 @@ def _has_running_concurrent_job(self, errors: Optional[Iterable[Mapping[str, Any
+        1) `aliased` or `hidden` store names from being exposed
+        2) migrated to data to the new store, but referenced within the old one stil.
+
+        reference issue: https://github.com/airbytehq/oncall/issues/5866
```
⭐

brianjlai commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:
```diff
@@ -283,14 +325,19 @@ def _on_job_with_errors(self, errors: List[Mapping[str, Any]]) -> AirbyteTracedE
-        raise ShopifyBulkExceptions.BulkJobNonHandableError(f"The Stream: `{self.stream_name}`, Non-handable error occured: {errors}")
+        raise ShopifyBulkExceptions.BulkJobNonHandableError(f"The Stream: `{self.http_client._name}`, Non-handable error occured: {errors}")
+
+    def _get_server_errors(self, response: requests.Response) -> List[Optional[dict]]:
```
nit:  for consistency can we make this `Mapping[str, Any]`. We'll also end up changing `_collect_bulk_errors()` signature, but that probably should be done anyway since the consuming method `_job_healthcheck` also uses a mapping

bazarnov replied:
Corrected! Thanks!

bazarnov commented:
New `pre-release` tag has been published: `2.4.15-dev.e7209419cc`

- Contains corrections and more information for the logs.

bazarnov commented:
New `pre-release` tag has been published: `2.4.15-dev.7162a62fbf`

- Inf loop fix, release candidate.

brianjlai commented on airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/job.py:
```diff
@@ -339,6 +402,32 @@ def _has_running_concurrent_job(self, errors: Optional[Iterable[Mapping[str, Any
+    def _switch_base_url(self) -> None:
+        if self._new_base_url:
+            self.base_url = self._new_base_url
+        else:
```
A lot of this flow to get the new shop name, raise `ShopifyBulkExceptions.BulkJobRedirectToOtherShopError`, catch the exception, and then switch the base URL if a new one exists feels pretty complex and confusing.

I'm trying to think if we can streamline this process a bit. I think raising the exception is fair because it is effectively a retry and we need to signal that the stream should perform a retry. Was there a reason why you opted to separate replacing the shop URL into two different methods that need to stay in sync w/ each other as shown by this `else` warning?

If we were to replace the URL in `_should_switch_shop_name()`, then we don't need to check if `_new_base_url` exists and the act of raising the exception to retry would assume that the URL was properly switched.

bazarnov replied:
Corrected! Thanks. 

The reason behind this, was to `inform` the customer `from` what we switch `to` what we switch, thus 2 vars were needed. I've managed to have 1 instead. Fixed.

brianjlai reviewed (approved):
Things look good, nice work on all these OC issues. One recommendation in the future would be to break these up into individual PRs. It was kind of hard to sort through things to understand what was being fixed for each OC issues especially since I'm not that familiar with the shopify code. But not a problem this time around.

I did add a few comments that we can discuss and also the `http_client.name` property should be in CDK `3.9.3` if you want to uptake that so we don't have to reference a private `_name` attribute.

And do we have any regression tests run against this new version of shopify?

bazarnov commented:
> And do we have any regression tests run against this new version of shopify?

Yes, @aldogonzalez8 helped with this, they are identical. No changes or regressions are expected.

aldogonzalez8 commented:
Acceptance testing is passing in combination with my PR branch.

Testing CAT steps to run in my local:
```
git checkout baz/source/shopify/fix-base-url-mismatch
git merge aldogonzalez8/fix_state_not_updating_for_stream_with_records_desc_order
airbyte-ci connectors --name=source-shopify --use-local-cdk test --only-step="acceptance"
```

![image](https://github.com/user-attachments/assets/e219e063-d6b7-4ce1-83ec-23fdef458d9d)


**Now, I will run a Regression locally**

aldogonzalez8 commented:
CAT results

```
airbyte-ci connectors --name=source-shopify --use-local-cdk build --architecture=linux/arm64

poetry run pytest src/live_tests/regression_tests --connector-image=airbyte/source-shopify --pr-url=https://github.com/airbytehq/airbyte/pull/42095 --connection-id=auto

...
run with state = yes
[?] Which connection would you like to retrieve?: 
 > https://cloud.airbyte.com/workspaces/112412bb-78aa-4ca6-aace-ae0b83be2215/connections/5f1d6da1-87bc-46b4-8de7-1b3fe1a2be90 - Internal: True - Stream count: 38 - Last attempt duration: 4 minutes
...
```

![image](https://github.com/user-attachments/assets/15675c3f-09b0-41dd-acaf-4bfcfdfd562e)
![image](https://github.com/user-attachments/assets/47272394-e97d-4e59-8544-925f549ca498)
![image](https://github.com/user-attachments/assets/13290f95-f2c7-4b95-9072-f3877c83d6ff)

![image](https://github.com/user-attachments/assets/61962295-4154-4998-9f2c-757deb02d0f3)

![image](https://github.com/user-attachments/assets/5373f4d1-90be-4344-a42e-b06fcadc8e07)
![image](https://github.com/user-attachments/assets/1c03400d-2769-4b21-919c-703f32dc0cf2)
![image](https://github.com/user-attachments/assets/11fe69db-d305-4c74-aaea-a60c7b7df21f)

![image](https://github.com/user-attachments/assets/0658e045-8ab6-48ec-81ff-cd7adbd2bdc0)

**product_variants example record for control version**
```

{
  "admin_graphql_api_id": "gid://shopify/ProductVariant/43315735396541",
  "available_for_sale": true,
  "barcode": "123",
  "compare_at_price": "1.00",
  "created_at": "2023-12-11T10:37:41+00:00",
  "display_name": "8 Ounce Soy Candle - Test Variant",
  "fulfillment_service": "manual",
  "grams": 127,
  "id": 43315735396541,
  "image_id": null,
  "inventory_item_id": 45419395743933,
  "inventory_management": "SHOPIFY",
  "inventory_policy": "CONTINUE",
  "inventory_quantity": 1,
  "old_inventory_quantity": 1,
  "options": [
    {
      "name": "Title",
      "option_value": {
        "has_variants": true,
        "id": "gid://shopify/ProductOptionValue/1646832976061",
        "name": "Test Variant",
        "swatch": null
      },
      "value": "Test Variant"
    }
  ],
  "position": 4,
  "presentment_prices": [
    {
      "compare_at_price": {
        "amount": 1.0,
        "currency_code": "USD"
      },
      "price": {
        "amount": 113.0,
        "currency_code": "USD"
      }
    }
  ],
  "price": 113.0,
  "product_id": 6796229509309,
  "requires_shipping": true,
  "shop_url": "airbyte-integration-test",
  "sku": "123",
  "tax_code": "",
  "taxable": true,
  "title": "Test Variant",
  "updated_at": "2023-12-11T10:37:41+00:00",
  "weight": 127.0,
  "weight_unit": "GRAMS"
}
```

**product_variants example record for target version**
```

{
  "admin_graphql_api_id": "gid://shopify/ProductVariant/43315735396541",
  "available_for_sale": true,
  "barcode": "123",
  "compare_at_price": "1.00",
  "created_at": "2023-12-11T10:37:41+00:00",
  "display_name": "8 Ounce Soy Candle - Test Variant",
  "fulfillment_service": "manual",
  "grams": 127,
  "id": 43315735396541,
  "image_id": null,
  "inventory_item_id": 45419395743933,
  "inventory_management": "SHOPIFY",
  "inventory_policy": "CONTINUE",
  "inventory_quantity": 1,
  "old_inventory_quantity": 1,
  "options": [
    {
      "name": "Title",
      "option_value": {
        "has_variants": true,
        "id": "gid://shopify/ProductOptionValue/1646832976061",
        "name": "Test Variant",
        "swatch": null
      },
      "value": "Test Variant"
    }
  ],
  "position": 4,
  "presentment_prices": [],
  "price": 113.0,
  "product_id": 6796229509309,
  "requires_shipping": true,
  "shop_url": "airbyte-integration-test",
  "sku": "123",
  "tax_code": "",
  "taxable": true,
  "title": "Test Variant",
  "updated_at": "2023-12-11T10:37:41+00:00",
  "weight": 127.0,
  "weight_unit": "GRAMS"
}
```

![image](https://github.com/user-attachments/assets/994d28c0-a523-45a3-9829-057b2967157a)

![image](https://github.com/user-attachments/assets/d7ea38cf-2937-4aab-91d4-12806146151a)

aldogonzalez8 commented:
Could this one be an issue? @bazarnov 
![image](https://github.com/user-attachments/assets/039ccbc2-216c-40cc-a5ba-59e9ed33d147)

bazarnov commented:
@aldogonzalez8 It was a question of the new config property:
- `job_product_variants_include_pres_prices: true`, which was missing for the `config` used for the latest `regression / live tests`. Fixed and updated the branch. 

I've checked the data locally, and the `presentment_prices` and the value is there, after falling back to the default behavior, for this record id `43315735396541` :
```
"presentment_prices":[{"price":{"amount":113.0,"currency_code":"USD"},"compare_at_price":{"amount":1.0,"currency_code":"USD"}}]
```

We can repeat the tests, if needed.

aldogonzalez8 commented:
Test repeated from scratch,  everything is 1:1

```
airbyte-ci connectors --name=source-shopify --use-local-cdk build --architecture=linux/arm64

poetry run pytest src/live_tests/regression_tests --connector-image=airbyte/source-shopify --pr-url=https://github.com/airbytehq/airbyte/pull/42095 --connection-id=auto

...
run with state = yes
[?] Which connection would you like to retrieve?: 
 > https://cloud.airbyte.com/workspaces/112412bb-78aa-4ca6-aace-ae0b83be2215/connections/5f1d6da1-87bc-46b4-8de7-1b3fe1a2be90 - Internal: True - Stream count: 38 - Last attempt duration: 4 minutes
...
```

![image](https://github.com/user-attachments/assets/328241d4-fe15-44ec-8d55-a36eada8bfe8)
![image](https://github.com/user-attachments/assets/a0798120-ade9-469b-9a16-6e31d1a9cf54)
![image](https://github.com/user-attachments/assets/c59f1fc0-9111-4088-a06e-770e85594923)
![image](https://github.com/user-attachments/assets/73a1d6d9-e425-485a-9897-6941e5852a3e)
![image](https://github.com/user-attachments/assets/ab4a744c-5536-4df0-8e25-2ce9ea4ec57c)
![image](https://github.com/user-attachments/assets/b47fc328-9698-4006-b46a-b9308205799a)

aldogonzalez8 commented:
/approve-regression-tests

bazarnov commented:
It will be released on Monday, July 29th.

aldogonzalez8 commented:
/approve-regression-tests

jblakeman commented on airbyte-integrations/connectors/source-shopify/source_shopify/spec.json:
```diff
@@ -98,12 +98,27 @@
-        "default": 3600,
-        "minimum": 1
+        "default": 7200,
+        "minimum": 3600,
```
Curious, is there a specific reason the minimum needs to be 3600 or greater than 1? Adding this requirement means large jobs significantly increase the time to which they arrive at a small enough slice that can run without failure.

bazarnov replied:
@jblakeman Thank you for raising this; we selected the values from what we got on our side and are happy to adjust them now that we have your feedback. I'll add this to the next source-shopify update.
